fdbcli/RangeConfigCommand.actor.cpp

```diff
@@ -0,0 +1,144 @@
+#include "fdbclient/NativeAPI.actor.h"
+#include "fdbclient/DataDistributionConfig.actor.h"
+
+#include "flow/actorcompiler.h" // This must be the last #include.
+
+namespace fdb_cli {
+
+ACTOR Future<bool> rangeConfigCommandActor(Database cx, std::vector<StringRef> tokens) {
+	state std::function<bool(std::string)> fail = [&](std::string msg) {
+		if (!msg.empty()) {
+			fmt::print(stderr, "ERROR: {}\n", msg);
+		}
+		printUsage(tokens[0]);
+		return false;
+	};
+
+	state std::list<StringRef> args(tokens.begin() + 1, tokens.end());
+	state std::function<StringRef()> nextArg = [&]() {
+		ASSERT(!args.empty());
+		auto s = args.front();
+		args.pop_front();
+		return s;
+	};
+	state std::function<int()> nextArgInt = [&]() {
+		if (args.empty()) {
+			throw boost::bad_lexical_cast();
+		}
+		return boost::lexical_cast<int>(nextArg().toString());
+	};
+
+	if (args.size() < 1) {
+		return fail("No subcommand given.");
+	}
+
+	state StringRef cmd = nextArg();
+
+	if (cmd == "show"_sr) {
+		state bool includeDefault = false;
+
+		while (!args.empty()) {
+			auto arg = nextArg();
+			if (arg == "includeDefault"_sr) {
+				includeDefault = true;
+			} else {
+				return fail(fmt::format("Unknown argument: '{}'", arg.printable()));
+			}
+		}
+
+		DDConfiguration::RangeConfigMapSnapshot config = wait(DDConfiguration().userRangeConfig().getSnapshot(
+		    SystemDBWriteLockedNow(cx.getReference()), allKeys.begin, allKeys.end));
+		fmt::print(
+		    "{}\n",
+		    json_spirit::write_string(DDConfiguration::toJSON(config, includeDefault), json_spirit::pretty_print));
+
+	} else if (cmd == "update"_sr || cmd == "set"_sr) {
+		if (args.size() < 3) {
+			return fail("Begin, end, and at least one configuration option are required.");
+		}
+
+		state KeyRef begin = nextArg();
+		state KeyRef end = nextArg();
+		if (end <= begin) {
+			return fail("Range end must be > range begin.");
+		}
+
+		state DDRangeConfig rangeConfig;
+
+		while (!args.empty()) {
+			state StringRef option = nextArg();
+
+			try {
+				if (option == "replication"_sr) {
+					rangeConfig.replicationFactor = nextArgInt();
+				} else if (option == "teamID"_sr) {
+					rangeConfig.teamID = nextArgInt();
+				} else if (option == "default"_sr) {
+					rangeConfig = DDRangeConfig();
+				} else {
+					return fail(fmt::format("Unknown range option: '{}'", option.printable()));
+				}
+			} catch (...) {
+				return fail(
+				    fmt::format("Required argument for range option '{}' missing or invalid.", option.toString()));
+			}
+
+			wait(DDConfiguration().userRangeConfig().updateRange(
+			    SystemDBWriteLockedNow(cx.getReference()), begin, end, rangeConfig, cmd == "set"_sr));
+		}
+	} else {
+		return fail(fmt::format("Unknown command: '{}'", cmd.printable()));
+	}
+
+	return true;
+}
+
+CommandFactory rangeConfigFactory(
+    "rangeconfig",
+    CommandHelp(
+        "rangeconfig show [includeDefault] | (update|set) <beginKey> <endKey> [default] [replication <N>] [teamID <N>]",
```

Comment:
This command factory takes two additional parameters for generating hints and auto-complete options for fdbcli commands. You can see some examples of this in TenantCommands.actor.cpp, MetaclusterCommands.actor.cpp, ThrottleCommands.actor.cpp, etc.

Reply:
Ah okay thanks, yeah I noticed that the CLI was not auto-completing past the command itself and didn't know if it was capable of doing that.

Reply:
Will that work on 71.2 as well?

Reply:
I'll defer this for later.  It's not immediately obvious to me what completion vs hints are or how this framework works exactly / what it should return and why, and I don't see any comments about it at

https://github.com/apple/foundationdb/blob/4a57aa6e4c5090c432323d8dd079850739cb9f8a/fdbcli/include/fdbcli/fdbcli.actor.h#L64

Reply:
I'd recommend doing at least the hint text now, as without it the implementation of this is somewhat incomplete. The implementation is not documented but is somewhat straightforward from following examples. The autocomplete generator is pretty valuable too, and I don't think it would be a significant effort to add that here either.